elkscmd/basic/host-ibmpc.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Architecture Specific routines for CGA
+ * Sep 2024 Takahiro Yamada
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#if __ia16__
+#include <arch/io.h>
+#else
+#define inb(port)           0
+#define inw(port)           0
+#define outb(value,port)
+#define outw(value,port)
+#endif
+#include "host.h"
+#include "basic.h"
+
+#define VIDEO_05_G320x200   0x0005
+#define VIDEO_03_T80x25     0x0003
+
+extern FILE *outfile;
+
+static int gmode = 0;
+static int exit_on = 0;
+
+typedef struct {
+    int x;
+    int y;
+    int fgc;
+    int bgc;
+    int r;
+} xyc_t;
+
+static xyc_t gxyc = {0, 0, 7, 0, 1};
+
+#if __ia16__
+void int_10(unsigned int ax, unsigned int bx,
+            unsigned int cx, unsigned int dx)
+{
+    __asm__ volatile ("push %ds;"
+                      "push %es;"
+                      "push %bp;"
+                      "push %si;"
+                      "push %di;");
+    __asm__ volatile ("int $0x10;"
+                      :
+                      :"a" (ax), "b" (bx), "c" (cx), "d" (dx)
+                      :"memory", "cc");
+    __asm__ volatile ("pop %di;"
+                      "pop %si;"
+                      "pop %bp;"
+                      "pop %es;"
+                      "pop %ds;");
+}
+
+void memclrw(unsigned int offset, unsigned int seg, unsigned int count)
+{
+    __asm__ volatile ("push %ds;"
+                      "push %es;"
+                      "push %bp;"
+                      "push %si;"
+                      "push %di;");
```

Comment:
Contrary to above, in this function obviously ES and ES will have to be saved explicitly. But the compiler should generate the normal function prologue, see below.